src/Biogeochemistry.jl

```diff
@@ -0,0 +1,134 @@
+abstract type AbstractBiogeochemistry end
+
+@inline (::AbstractBiogeochemistry)(i, j, k, grid, val_tracer_name, clock, fields) = zero(grid)
+
+struct NoBiogeochemistry <: AbstractBiogeochemistry end
+
+tracernames(tracers) = keys(tracers)
+tracernames(tracers::Tuple) = tracers
+
+"""Ensure that `tracers` contains biogeochemical tracers and `auxiliary_fields` contains biogeochemical auxiliary fields (e.g. PAR)."""
+@inline function validate_biogeochemistry!(tracers, auxiliary_fields, bgc)
+    req_tracers = required_biogeochemical_tracers(bgc)
+    
+    all(tracer ∈ tracernames(tracers) for tracer in req_tracers) ||
+        error("$(req_tracers) must be among the list of tracers to use $(typeof(bgc).name.wrapper)")
+
+    req_auxiliary_fields = required_biogeochemical_auxiliary_fields(bgc)
+    
+    all(field ∈ tracernames(auxiliary_fields) for field in req_auxiliary_fields) ||
+        error("$(req_auxiliary_fields) must be among the list of auxiliary fields to use $(typeof(bgc).name.wrapper)")
+    
+    return nothing
+end
+
+required_biogeochemical_tracers(::NoBiogeochemistry) = ()
+required_biogeochemical_auxiliary_fields(bgc::AbstractBiogeochemistry) = ()
+
+"""Sets up a tracer based biogeochemical model in a similar way to SeawaterBouyancy"""
+struct TracerBasedBiogeochemistry <: AbstractBiogeochemistry
+    biogeochemical_tracers
+    reactions
+    advection_scheme
+    sinking_velocities
+    auxiliary_fields
+end
+
+function regularize_sinking_velocities(sinking_speeds)
+    sinking_velocities = []
+    for w in values(sinking_speeds)
+        u, v, w = maybe_constant_field.((0.0, 0.0, - w))
+        push!(sinking_velocities, (; u, v, w))
+    end
+
+    return NamedTuple{keys(sinking_speeds)}(sinking_velocities)
+end
+
+# we can't use the standard `ContinuousForcing` regularisation here because it requires all the tracers to be inplace to have the correct indices
+struct ContinuousBiogeochemicalForcing
+    func
+    params
+    field_dependencies
+end
+
+DiscreteBiogeochemicalForcing = DiscreteForcing
+
+ContinuousBiogeochemicalForcing(func; parameters=nothing, field_dependencies=()) = ContinuousBiogeochemicalForcing(func, parameters, field_dependencies)
+
+function BiogeochemicalForcing(func; parameters=nothing, discrete_form=false, field_dependencies=())
+    if discrete_form
+        return DiscreteBiogeochemicalForcing(func, parameters)
+    else
+        return ContinuousBiogeochemicalForcing(func, parameters, field_dependencies)
+    end
+end
+
+function regularize_biogeochemical_forcing(forcing::Function)
+    return ContinuousBiogeochemicalForcing(forcing)
+end
+
+regularize_biogeochemical_forcing(forcing) = forcing
+
+@inline getargs(fields, field_dependencies, i, j, k, params::Nothing) =@inbounds map(field_name -> fields[field_name][i, j, k], field_dependencies)
+@inline getargs(fields, field_dependencies, i, j, k, params) = @inbounds tuple(map(field_name -> fields[field_name][i, j, k], field_dependencies)..., params)
```

Comment:
Doesn't this function exist elsewhere?

Reply:
Ah I should probably use these like continous forcing does:
https://github.com/CliMA/Oceananigans.jl/blob/0163628e3dc810eed76440cc6ceef9870cc50f7f/src/Operators/interpolation_utils.jl#L34-L42
But I don't really understand why these are defined 6 times?